pubtools/_pulp/tasks/common.py

```diff
@@ -179,8 +182,9 @@ def publish_with_cache_flush(
         # hook implementation(s) may now flush pulp-derived caches and datastores
         pm.hook.task_pulp_flush()
 
-        # flush CDN cache
-        out = self.flush_cdn(repos)
+        # flush cdn cache if publish is not being skipped
+        if "publish" not in self.args.skip:
```

Comment:
@rohanpm i'm trying to figure out how the skip argument might be accessed from within the Publisher class in common.py but it seems to be proving difficult looking for some advice. Currently tests are failing like so for publish/:
```
    def publish_with_cache_flush(
        self, repos, units=None, pulp_client=None, errata=None
    ):
        # Ensure all repos in 'repos' are fully published, and CDN/UD caches are flushed.
        #
        # If 'units' are provided, ensures those units have cdn_published field set after
        # the publish and before the UD cache flush.
        #
        # Initialize out variable as an empty list in the case where publish and cdn
        # cache flush are skipped
        units = units or []
        pulp_client = pulp_client or self.pulp_client
        out = []
    
        # publish the repos found
        publish_fs = self.publish(repos)
    
        # wait for the publish to complete before
        # flushing caches.
        f_sequence(publish_fs).result()
    
        # hook implementation(s) may now flush pulp-derived caches and datastores
        pm.hook.task_pulp_flush()
    
        # flush cdn cache if publish is not being skipped
>       if "publish" not in self.args.skip:
E       AttributeError: 'Namespace' object has no attribute 'skip'

pubtools/_pulp/tasks/common.py:186: AttributeError
```
i have been trying to replicate what is seen here at https://github.com/release-engineering/pubtools-pulp/blob/master/pubtools/_pulp/tasks/push/command.py#L51 to be used in a similar fashion in this Publisher class

any advice is appreciated,
Thanks

Reply:
@lekang9 

That failure seems to be in test_fix_cves.  The problem is that the argument parser used by the fix-cves command doesn't add a `--skip` argument.  So you can't read it.

Your choices would seem to be:

- Make the fix-cves command also accept `--skip` (I like this option, I would like it to be standard for all commands).  This would be done by adding it to `FixCves.add_args` method.
- Or, change the code here reading `skip` to accept when the attribute is missing, e.g. by using `getattr` rather than accessing it directly.

Reply:
Thanks i ended up going with moving it into PulpTask._basic_args which helped to pass the publish tests, however it looks like the changes i made to clearrepo are still causing failures i'll keep looking into that

Reply:
@rohanpm i have another question unrelated to above, so i've cleared majority of the tests however with the changes to clear_repo.py i am still experiencing some failures due to TypeError where f_flat_map expects a future but this is not an issue where it is used elsewhere, my first suspicion is that the tests for clear_repo need to be adjusted to be more similar to test_publish. I'll push a commit shortly with what i have so far so that the tests update here

Reply:
@lekang9 , evidently `publish` (and by extension `publish_with_cache_flush`) are designed to accept a `list[Future[Repository]]` and you've changed the code here to now pass it a `list[Repository]` instead.  It might help if the code had type hints, but it wasn't an option when the code was originally written.

Note if you're not sure about the type of things, you can run py.test with `--pdb` argument and it'll drop you into a debug session immediately when a test fails, and you can then inspect various objects.

In this case, in order to call `publish_with_cache_flush` with a `list[Future[Repository]]` as it's designed for, you could write:

```
         # They should be published.
-        publish_fs = self.publish_with_cache_flush(repos)
+        publish_fs = self.publish_with_cache_flush([f_map(f, lambda cr: cr.repo) for f in cleared_repos_fs])
```

Reply:
Thanks that is helpful for working through these tests, i've hit another snag in the tests but this is more of a question specific to clear_repo and whether or not we should be having step "set_cdn_published" included because in the diff seen previously we are not doing so (even when flushing ud cache) imo it seems to make sense. Atleast from seeing the comments in the code it makes sense to have this step before flushing ud cache in general.
here is a diff after i've applied the f_map to publish_with_cache_flush in clear_repo.py
```
>           raise AssertionError(message)
E           AssertionError: Output differs from expected (set UPDATE_BASELINES=1 if intended):
E           --- <output of test>
E           +++ /home/lekang/Repos/pubtools-pulp/tests/logs/clear_repo/test_clear_repo/test_clear_repo_multiple_content_types.txt
E           @@ -7,10 +7,12 @@
E            [    INFO] Record push items: finished
E            [    INFO] Publish: started
E            [    INFO] Publish: finished
E           +[    INFO] Flush CDN cache: started
E           +[    INFO] CDN cache flush is not enabled.
E           +[    INFO] Flush CDN cache: finished
E           +[    INFO] Set cdn_published: started
E           +[    INFO] Set cdn_published: finished
E            [    INFO] Flush UD cache: started
E            [    INFO] UD cache flush is not enabled.
E            [    INFO] Flush UD cache: finished
E           -[    INFO] Flush CDN cache: started
E           -[    INFO] CDN cache flush is not enabled.
E           -[    INFO] Flush CDN cache: finished
```

Reply:
FWIW i've updated baselines but there are 2 tests still failing it looks like there is an issue with the url to be flushed not making it through somewhere e.g.
```
        # It should have flushed these URLs
>       assert sorted(task_instance.fastpurge_client.purged_urls) == [
            "https://cdn.example.com/some/publish/url/mutable1",
            "https://cdn.example.com/some/publish/url/mutable2",
        ]
E       AssertionError: assert [] == ['https://cdn...url/mutable2']
E         
E         Right contains 2 more items, first extra item: 'https://cdn.example.com/some/publish/url/mutable1'
E         
E         Full diff:
E         + []
E         - [
E         -     'https://cdn.example.com/some/publish/url/mutable1',
E         -     'https://cdn.example.com/some/publish/url/mutable2',
E         - ]

tests/clear_repo/test_clear_repo.py:197: AssertionError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/lekang/Repos/pubtools-pulp/tests/clear_repo/test_clear_repo.py(197)test_clear_file_repo()
-> assert sorted(task_instance.fastpurge_client.purged_urls) == [
(Pdb) print(task_instance.fastpurge_client.purged_urls
*** SyntaxError: unexpected EOF while parsing
(Pdb) print(task_instance.fastpurge_client.purged_urls)
[]
```
i can only assume that with my input the relative url to be flushed is being lost somewhere along the lines

Reply:
> Thanks that is helpful for working through these tests, i've hit another snag in the tests but this is more of a question specific to clear_repo and whether or not we should be having step "set_cdn_published" included because in the diff seen previously we are not doing so (even when flushing ud cache) imo it seems to make sense. 

`set_cdn_published` needs to be called with some units, so I'm not sure what units you could pass in the clear_repo case.  The ones just removed from the repo wouldn't make sense as they're not being published to the CDN.

I don't think it can be *meaningfully* invoked in the clear repo case, but if the method is just getting called, logging those lines but not actually doing anything, I think that's fine.

Reply:
> FWIW i've updated baselines but there are 2 tests still failing it looks like there is an issue with the url to be flushed not making it through somewhere e.g.

I can help you with this if you push the updated code, but I'm only seeing code from 4 days ago.

Reply:
Sorry i forgot to update it, i've just pushed it up

Reply:
@lekang9 

Did you notice that the baseline update for test_clear_file_repo.txt included an error?

```
--- a/tests/logs/clear_repo/test_clear_repo/test_clear_file_repo.txt
+++ b/tests/logs/clear_repo/test_clear_repo/test_clear_file_repo.txt
@@ -7,10 +7,6 @@
 [    INFO] Record push items: finished
 [    INFO] Publish: started
 [    INFO] Publish: finished
-[    INFO] Flush UD cache: started
-[    INFO] Flush UD cache: finished
 [    INFO] Flush CDN cache: started
-[    INFO] Flushing cache for some-filerepo:
-[    INFO]    https://cdn.example.com/some/publish/url/mutable1
-[    INFO]    https://cdn.example.com/some/publish/url/mutable2
-[    INFO] Flush CDN cache: finished
+[   ERROR] Flush CDN cache: failed
+# Raised: 'MapFuture' object has no attribute 'relative_url'
```

That exception is what needs to be resolved first, don't worry about the assertion.

The error can also be seen if you run the test with `--log-cli-level=DEBUG`

```
$ py.test -xv tests/clear_repo/test_clear_repo.py::test_clear_file_repo --log-cli-level=DEBUG
...
-------------------------------------------- Captured stderr call ---------------------------------------------
Traceback (most recent call last):
  File "/home/rmcgover/src/pubtools-pulp/tests/command.py", line 91, in test
    fn()
  File "/home/rmcgover/src/pubtools-pulp/pubtools/_pulp/task.py", line 164, in main
    self.run()
  File "/home/rmcgover/src/pubtools-pulp/pubtools/_pulp/tasks/clear_repo.py", line 135, in run
    publish_fs = self.publish_with_cache_flush(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rmcgover/src/pubtools-pulp/pubtools/_pulp/tasks/common.py", line 187, in publish_with_cache_flush
    out = self.flush_cdn(repos)
          ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rmcgover/src/pubtools-pulp/pubtools/_pulp/step.py", line 37, in new_fn
    ret = fn(instance, *args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rmcgover/src/pubtools-pulp/pubtools/_pulp/tasks/common.py", line 78, in flush_cdn
    return [purge_repo(r) for r in repos if r.relative_url]
                                            ^^^^^^^^^^^^^^
AttributeError: 'MapFuture' object has no attribute 'relative_url'
```

This is a baseline update that you should not have accepted.
To update the baselines, you run the tests with `UPDATE_BASELINES=1`, review the result and then add it to your commit only if the new result looks right, which it doesn't in this case.

The root cause is type confusion again.  Unfortunately it looks like there's quite a mess here now.  I see that there's two `publish` methods, one expects a `list[Repository]` and the other a `list[Future[Repository]]`.  This particular code expects to be seeing `Repository` and not `Future[Repository]`, but we already determined that other nearby code paths are expecting to see futures.

So: I think this code has been inadvertently written in such a way that it only works if the given objects can be treated as *both* `Future[Repository]` and `Repository`.

That sucks, but anyway an easy fix would be:

```
--- a/pubtools/_pulp/tasks/clear_repo.py
+++ b/pubtools/_pulp/tasks/clear_repo.py
@@ -2,7 +2,7 @@ import logging
 from functools import partial
 
 import attr
-from more_executors.futures import f_map, f_sequence
+from more_executors.futures import f_map, f_proxy, f_sequence
 from pubtools.pulplib import ContainerImageRepository, Criteria
 
 from pubtools._pulp.arguments import SplitAndExtend
@@ -133,7 +133,7 @@ class ClearRepo(
 
         # They should be published.
         publish_fs = self.publish_with_cache_flush(
-            [f_map(f, lambda cr: cr.repo) for f in cleared_repos_fs]
+            [f_proxy(f_map(f, lambda cr: cr.repo)) for f in cleared_repos_fs]
         )
 
         # Wait for everything to finish
```

...which allows the code to continue being sloppy.

I think it'd be better if all this code was reviewed and type hints added to resolve the confusion throughout, but I don't think we can expect that to fall into the scope of your current issue.

Reply:
yes i did notice that there was a separate error but i could not figure out why, it just felt that the inputs were not right but you have answered it as above

Reply:
i'll go ahead and revert baseline update from the last commit and then update it to use f_proxy

Reply:
so i've reverted the previous commit and changed the input to what you gave me above and there was only baseline updates related to "set_cdn_published" which looks alot better so i went ahead and did that but i think that the skipping log can be handled better:
```
>           raise AssertionError(message)
E           AssertionError: Output differs from expected (set UPDATE_BASELINES=1 if intended):
E           --- <output of test>
E           +++ /home/lekang/Repos/pubtools-pulp/tests/logs/clear_repo/test_clear_repo/test_clear_file_skip_publish.txt
E           @@ -6,10 +6,9 @@
E            [    INFO] Record push items: started
E            [    INFO] Record push items: finished
E            [    INFO] Publish: skipped
E           +[    INFO] Set cdn_published: started
E           +[    INFO] Set cdn_published: finished
E            [    INFO] Flush UD cache: started
E            [    INFO] UD cache flush is not enabled.
E            [    INFO] Flush UD cache: finished
E           -[    INFO] Flush CDN cache: started
E           -[    INFO] CDN cache flush is not enabled.
E           -[    INFO] Flush CDN cache: finished
```
in relation to how flushing of cdn cache is flushed when publish is skipped, its simply skipped but there is no info logging there. going to look into that i think i should make proper use of the step decorator to skip the flush_cdn step so that it does what it would for publish (if that makes sense)

Reply:
This looks better then there being no logs
```
>           raise AssertionError(message)
E           AssertionError: Output differs from expected (set UPDATE_BASELINES=1 if intended):
E           --- <output of test>
E           +++ /home/lekang/Repos/pubtools-pulp/tests/logs/clear_repo/test_clear_repo/test_clear_file_skip_publish.txt
E           @@ -6,6 +6,9 @@
E            [    INFO] Record push items: started
E            [    INFO] Record push items: finished
E            [    INFO] Publish: skipped
E           +[    INFO] Flush CDN cache: started
E           +[    INFO] CDN cache flush is not enabled.
E           +[    INFO] Flush CDN cache: finished
E            [    INFO] Set cdn_published: started
E            [    INFO] Set cdn_published: finished
E            [    INFO] Flush UD cache: started

tests/command.py:208: AssertionError
```

Reply:
I tried to address some of the complaints in static that seemed simpler to address
```
static create: /home/lekang/Repos/pubtools-pulp/.tox/static
static installdeps: -rtest-requirements.txt
static inst: /home/lekang/Repos/pubtools-pulp/.tox/.tmp/package/1/pubtools-pulp-1.27.1.zip
static installed: alabaster==0.7.13,astroid==3.0.1,attrs==23.1.0,Babel==2.13.1,bandit==1.7.5,black==23.11.0,certifi==2023.11.17,cffi==1.16.0,cfgv==3.4.0,charset-normalizer==3.3.2,click==8.1.7,coverage==6.5.0,coveralls==3.3.1,cryptography==41.0.5,decorator==5.1.1,dill==0.3.7,distlib==0.3.7,docopt==0.6.2,docutils==0.20.1,edgegrid-python==1.3.1,exceptiongroup==1.2.0,fastpurge==1.0.4,filelock==3.13.1,frozendict==2.3.9,frozenlist2==1.0.0,gitdb==4.0.11,GitPython==3.1.40,gssapi==1.8.3,humanize==4.9.0,identify==2.5.32,idna==3.6,imagesize==1.4.1,importlib-metadata==6.8.0,importlib-resources==6.1.1,iniconfig==2.0.0,iniparse==0.5,isort==5.12.0,Jinja2==3.1.2,jsonschema==4.20.0,jsonschema-specifications==2023.11.1,kobo==0.34.0,koji==1.33.1,markdown-it-py==3.0.0,MarkupSafe==2.1.3,mccabe==0.7.0,mdurl==0.1.2,mock==5.1.0,more-executors==2.11.4,mypy-extensions==1.0.0,ndg-httpsclient==0.5.1,nodeenv==1.8.0,packaging==23.2,pathspec==0.11.2,pbr==6.0.0,pkgutil_resolve_name==1.3.10,platformdirs==4.0.0,pluggy==1.3.0,pre-commit==3.5.0,pubtools==1.3.0,pubtools-pulp @ file:///home/lekang/Repos/pubtools-pulp/.tox/.tmp/package/1/pubtools-pulp-1.27.1.zip,pubtools-pulplib==2.38.0,pushcollector==1.3.0,pushsource==2.40.0,pyasn1==0.5.1,pycparser==2.21,Pygments==2.17.2,pylint==3.0.2,pyOpenSSL==23.3.0,pytest==7.4.3,pytest-cov==4.1.0,pytest-threadleak==0.5.0,python-dateutil==2.8.2,pytz==2023.3.post1,PyYAML==6.0.1,referencing==0.31.0,requests==2.31.0,requests-gssapi==1.2.3,requests-mock==1.11.0,requests-toolbelt==1.0.0,rhsm==1.19.2,rich==13.7.0,rpds-py==0.13.1,rpm==4.17.1,rpm-py-installer==1.2.0,six==1.16.0,smmap==5.0.1,snowballstemmer==2.2.0,Sphinx==7.1.2,sphinx-argparse==0.4.0,sphinxcontrib-applehelp==1.0.4,sphinxcontrib-devhelp==1.0.2,sphinxcontrib-htmlhelp==2.0.1,sphinxcontrib-jsmath==1.0.1,sphinxcontrib-qthelp==1.0.3,sphinxcontrib-serializinghtml==1.1.5,stevedore==5.1.0,tomli==2.0.1,tomlkit==0.12.3,typing_extensions==4.8.0,urllib3==2.1.0,virtualenv==20.24.7,zipp==3.17.0
static run-test-pre: PYTHONHASHSEED='2065978990'
static run-test: commands[0] | sh -c 'pylint pubtools; test $(( $? & (1|2|4|32) )) = 0'
************* Module pubtools._pulp.tasks.copy_repo
pubtools/_pulp/tasks/copy_repo.py:52:37: R1729: Use a generator instead 'any(not r_id.strip() for r_id in pair)' (use-a-generator)
************* Module pubtools._pulp.tasks.common
pubtools/_pulp/tasks/common.py:210:4: W0246: Useless parent or super() delegation in method 'add_args' (useless-parent-delegation)
************* Module pubtools._pulp.tasks.delete
pubtools/_pulp/tasks/delete.py:405:21: R1721: Unnecessary use of a comprehension, use list(adv_f.result()) instead. (unnecessary-comprehension)
************* Module pubtools._pulp.tasks.fix_cves
pubtools/_pulp/tasks/fix_cves.py:36:18: R1721: Unnecessary use of a comprehension, use list(erratum_page) instead. (unnecessary-comprehension)
pubtools/_pulp/tasks/fix_cves.py:92:15: R1721: Unnecessary use of a comprehension, use list(resp.result()) instead. (unnecessary-comprehension)
************* Module pubtools._pulp.tasks.push.items.base
pubtools/_pulp/tasks/push/items/base.py:126:23: R1718: Consider using a set comprehension (consider-using-set-comprehension)
************* Module pubtools._pulp.tasks.set_maintenance.set_maintenance_on
pubtools/_pulp/tasks/set_maintenance/set_maintenance_on.py:40:26: R1718: Consider using a set comprehension (consider-using-set-comprehension)

------------------------------------------------------------------
Your code has been rated at 9.98/10 (previous run: 9.94/10, +0.03)

ERROR: InvocationError for command /usr/bin/sh -c 'pylint pubtools; test $(( $? & (1|2|4|32) )) = 0' (exited with code 1)
```
i'm not sure why it still gives an error does it do so until all complaints are addressed?

Reply:
Rajul helped to point out which specific complaint was causing static to fail, it was confusing because it started with a W i assumed it was a warning rather than an error to fail on